mongo_connector/doc_managers/neo4j_doc_manager.py

```diff
@@ -53,13 +58,53 @@ def upsert(self, doc, namespace, timestamp):
         "ns": namespace,
         "_ts": timestamp
     }
+    doc = self._formatter.format_document(doc)
+    self.store_nodes_and_relationships(doc, doc_type, doc_id)
+    
   
+  def store_nodes_and_relationships(self, doc, doc_type, doc_id):
+    self.build_nodes(doc_type, doc, doc_id)
+    for node_type in self.doc_types:
+      self.apply_id_constraint(node_type)
+    tx = self.graph.cypher.begin()
+    for query in self.query_nodes.keys():
+      tx.append(query, {"parameters":self.query_nodes[query]})
+    main_type = self.doc_types.pop(0)
+    for node_type in self.doc_types:
+      tx.append(self.build_relationships(doc_id, main_type, node_type))
+    tx.commit()
+
+  def build_nodes(self, doc_type, hash, id):
+    self.doc_types.append(doc_type)
+    parameters = {'id':id}
+    for key in hash.keys():
+      if (type(hash[key]) is str) or (type(hash[key]) is list):
+        value = hash[key]
+        parameters.update({ key: value })
+      else:
+        self.build_nodes(key, hash[key], id)
+    query = "CREATE (c:Document:{doc_type} {{parameters}})".format(doc_type=doc_type)
+    self.query_nodes.update({query: parameters})
+
+  def build_relationships(self, doc_id, main_type, node_type):
+    relationship_type = main_type + "_" + node_type
+    statement = "MATCH (a:{main_type}), (b:{node_type}) WHERE a.id = '{doc_id}' AND b.id = '{doc_id}' CREATE (a)-[r:{relationship_type}]->(b)".format(main_type=main_type, node_type=node_type, doc_id=doc_id, relationship_type=relationship_type)
```

Comment:
Is it guaranteed that both ends (both nodes) already exist in the db when we create the rels? Otherwise this operation will silently not do anything as it won't find the nodes.
If we can't guarantee that we will have to use MERGE instead of MATCH but that's of course a more expensive operation.